### ISSUE TITLE
Unguided Campaign Discussion

### ISSUE THREAD

**Issue by Moburma** [mefistotelis/swars#58]
An issue to discuss and plan on making an Unguided campaign. At the moment the focus should be on getting levels together and fixing what is there, with an eye to making new levels (will need proper tooling for that). Also just general ideas what the campaign should be about/how it should play compared to the existing ones. I recommend you read this [this](https://tcrf.net/Syndicate_Wars_(DOS)/Unguided_Campaign) to get an overview of levels (full disclosure - I wrote it). Also if you want to see the existing levels being played to completion, check out my YouTube channel [here](https://www.youtube.com/watch?v=Nfml3yYqqzU&list=PLhFEaJK1kM7UNyTVyYrk-2EO4HyCyMhzw).

**Unique Selling Point**

What is the Unguided's USP - why bother with an Unguided campaign? I think this is important to consider beyond simply "more levels is cool!" The Unguided themselves are not that interesting from a gameplay point of view - they have less health and are weaker, and don't have special unique weaponry. From the coverage in Maximum magazine, it suggests that the Unguided campaign was to be an expert mode unlocked by beating one of the two other main campaigns first. It also suggests there would not be normal progression in this mode, something heavily backed up by the level load outs (limited numbers of units and usually basic weapons). My vision of this would be that the campaign if possible (and there is a technical limit here, the other campaigns do not work this way) should NOT have research and inventory screens, with that aspect of the game deliberately removed. Instead each level would have a set inventory the player must get the most out of. The advantages of this would be: a) more strategic gameplay. You can't just give your units the best guns anymore, forcing the player to learn new strategies and to use things like support weapons that otherwise come too late to be useful in the final game. b) allow the use of items in ways that would break the normal progression of the game. Want to give the player a Graviton Gun as a reward on the first level? You can now, without breaking the game. This makes things like the Chromotap and Displacertron actually viable and shows the player a side of the game they probably ignored in the main campaigns. c) Asymmetric weapon load outs. Putting the player into deliberately challenging and interesting situations they would not knowingly ever put themselves into.

In the game's campaigns, the Unguided have an anarchic trickster feel to them, using things like KO or Psycho Gas or Razorwire, this would also be interesting to play with. The limitation of weapons also has precedence in Bullfrog's games - Magic Carpet used this extensively, abandoning the traditional progression system for fixed loadouts as the game went on. Dungeon Keeper also used this to great effect by frequently limiting what skills, rooms and creatures the player could get between levels.

**Lore**

The game's plot states that essentially Eurocorp's elite scientific research team find alien artifacts. The leader of this group is Ko-Paull Vissick, who falls out with the rest over augmenting their brains with alien technology. This causes them to go insane, and him to sabotage their lab. They become the Nine, and he joins with the existing Unguided faction to become their leader. He seems to be some kind of bizarre quasi religious guru character, with Eurocorp questioning his sanity too (although what he says about the intentions of the zealots is proven to be 100% correct). It's obvious briefings would come from him, just not at all clear what his character would be. Mad religious leader (sounds a bit too similar to the zealots)? Cyberpunk hacker? Cerebral scientist?

From the final game Eurocorp briefings, the Unguided hole up in Delhi and attempt to flee to Tibet, where somehow they would survive the Zealot attack. Both locations also have mystical overtones, hinting at the guru/religious leader angle.

There are also obvious allusions to the manga/anime Akira with the biker gangs (there's even a suspiciously Kaneda-like NPC just to make it even more on the nose) and the apocalyptic religious cult rising up.

From a plot point of view there is a need to plan out what and where the Unguided and Vissick are said to do things to make any new campaign plot mesh with them. Can also be used for levels with reversed perspectives (e.g. you as Unguided defend _against_ Eurocorp on levels where you attacked them in the original campaign).

**Current Levels** 

In rough order I think they should go in, and my comments on bugs and how fun I think they are:

Index 68 Zealot Attack
Rating 3/5
This makes sense to be the first level, as it directly follows on from the start of the Zealot campaign. 
Very simple level (maybe too simple), but shows shortcomings of the follow command for the punk followers - a winning strategy is getting into the zealot's car, but if this is done your followers can't get in in and just get run over.
Complete Level

Index 80 Clean-Up
Rating 2/5
As an aside I don't think the level set here actually maps with description. Either way this is an extremely hard level that I think is actually impossible as-is. The basic idea - splitting up your group to cover two areas, and also defending civilians - is a really good one. Unfortunately provided equipment is insufficient. Not sure if this is actually salvagable as a level idea at all though, being frank.
Unfinished level

Level C007L007.D1
Rating 4/5
This is a loose level with no objectives set. A lot of effort has been put into it and it seems finished. It's seemingly a simple kill everything mission, but it's very challenging, which is fun and the kind of thing I think the campaign should be angled towards. The problem is it's really not clear what the objectives would be, and I think looking at the groups it's likely it was NOT simply kill everyone. The aim seems to be to get to the centre of the city, but it's not clear why, whether it's to kill the central groups or maybe persuade/escort someone there (I think this is very unlikely, but could be a fun objective). My hunch is this was actually the level called "Clean up" originally, as it matches the text description of the map. 
Finished level, but needs objectives/mission index entry created

Index 77 & 60 Guns Guns Guns
Rating 4/5 for 60, 3/5 for 77
Two versions of the same level. 60 is a mostly complete early version, but it seems this was rejected for a later much simpler version (77). That said the later version seems again impossibly hard, and also a bit boring in comparison to the early one. The later one could be interesting with some tweaks, it needs more weapons or something. The early version has bugs with some of the follower punks not joining the player, but I'm not sure if this is just a bug with my pre-alpha convertor, need to check again now early level support is in. The early version is also missing objectives, but it's quite obvious what they would be. Oddly killing the final scientist gives the player a sniper rifle, much like the level below - this could be a sign there WAS progression in the campaign.
Both levels seem unfinished and 60 needs objectives/mission index entry created

Index 64
Rating 4/5
Again there are two versions of this level. There is a mistake in the .miss file here and the wrong level is set. The objectives set in the file are for the level file c003l007.d1. This is a no brainer substitution, as this is a polished final game level, whereas the other version (which is wrongly set) is an early level file and nowhere near as good (and missing objectives). The final version also grants a sniper rifle for killing the scientist, like the level above, some kind of thematic idea between the two.
Finished level, needs .miss file corrected to point to level file c003l007.d1

Level C047L015.D1
Rating 2/5
Totally cut level that is quite polished but a little light on things to do do. That said there is quite a bit setup in this level, so could be stretched out to be more interesting. As it's totally cut has no mission objectives, but seems to just be an assassinate mission against Eurocorp.
Finished level, but needs objectives/mission index entry created

Level C004L007.DAT
Rating 3/5
Totally cut level. The objectives here seem quite obvious - there is a group of soldiers by a bank. Attacking them causes some of them to break off and flee in vehicles. They ride to the top of the freeway and head to a getaway flying car. This then flies to the edge of the map and then stops. Therefore it's a an assassinate mission with a twist to take out the baddies before they escape off the map. Works well, but it's a bit too simple once you understand what to do. There's also a problem that there seem to be no vacant cars, so the player can't get into an exciting chase with the soldiers, the only solution at the moment is to sprint to their cars and steal one, which ends things quickly in quite a dull manner. Also there is a big group of Zealots that could be interesting as a secondary objective.
Finished level, but needs objectives/mission index entry created

Index 78&61 Communications
Rating 3/5
Two versions of this one as well. I am pretty sure these are set to the wrong map file. Experimentation has shown that there doesn't actually seem to be a matching map for this level. If they are renamed to use map 79 Reykjavik, they just about work. If this is done they are quite interesting complex and very difficult (but doable) levels. Could be something really good, but needs a lot of work to make them playable.
Finished levels(?) but for a missing map, needs a lot of work to be coherent in the final game

Index 70
Rating 3/5
This is a simple and quite fun early level. Works perfectly, but could perhaps be improved by using flag1/2 objectives to draw the player's attention to the chemical plant where the essential grenade is (although a briefing could explain this as well).
Finished level

Index 66
Rating 3/5 
This is an early level that is a good idea, but seems a bit rough. Odd stuff like the player starting with a useless persuadertron - was the idea to gather the other unguided together? Feels like it needs something else added to really shine. There's an older version of this, but it's the same only clearly more unfinished.
Finished level - barely

Index 81 Godfather
Rating 3/5
Interesting and quite polished level, but not that much to it, which lets it down a little.
Finished level

Index 62 Rescue
Rating 3/5
This is a weird one. Doesn't make much sense to play, either needs a briefing to hint at what to actually do, or needs changing. Seems not quite finished.
Unfinished level

Index 82 Dumped
Rating 3/5
Another very hard level. Seems pretty finished, but either I'm missing something or it needs rebalancing as it's brutally hard (think it MIGHT be possible if you're really good/lucky though). Also seems to have had a wacky plot originally. The thing with the cars at the start blocking each other needs fixing.
Finished level

Index 65 Massive Assault
Rating 2/5
This seems good on paper, but it's actually too simplistic to really be fun. While it's marked as a late game level, some of the earlier .miss files have it as an early level (even the first!) and it seems to fit in better there. There's a bug where the wrong group is targetted that is trivial to fix, but it's just a boring easy kill everyone level. In the older files there are other objectives to get a weapon and take it to a location, but this is hardly thrilling stuff either. Either use as early filler, or needs radical rework.
Finshed level

Index 75 Ferocity
Rating 4/5
Important level that concludes the trilogy of similar levels in the other two campaigns. Feels a little rough in comparison, but works quite well.
Finished level

Index 67 Talks Ambush
Rating 4/5
I really like the idea of this one. It's very glitchy though, something about the collect item objectives is wrong and causes an overflow to random parts of the map. It's also clear that there is supposed to be some kind of counter-ambush from the unguided hiding in the buildings near the weapons cache, but they don't do anything, it seems this was never finished. 
Unfinished Level 

Index 69 Trash Church
Rating 1/5
An intriguing but barely even started level. Doesn't work in the final executable due to no player characters in the level. There is only even one enemy to kill. An all out war to blow up a zealot church sounds like an exciting idea for a level though.
Unfinished Level

Index 103 A Tale of Two Cities
Rating 5/5 
This was clearly going to be a demo for the Unguided campaign. Very polished and clever level full of surprises. Also lots of interesting stuff unused in the actual mission that could be spun off to another one.
Finished Level

**Next Steps**

My feelings are a) obviously there need to be more levels to be a proper campaign and b) a lot of these are not actually that great, they're clearly first drafts or unfinished and don't push the game's envelope much. 
The simplest first step forward would be to get objectives created for the totally cut levels, and try to fix the unfinished/broken ones. 
I think I will try and make a simple level editor like my Magic Carpet one, but my limited ability and the nature of this game means it's less suitable for that approach (really needs a realtime 3D view - someone with skills in e.g. Unity could make something out of [this project](https://github.com/RichDavisonNCL/UnitySWars).

**Comment by mefistotelis** [comment]
Some mission indexes currently have hard-coded properties. Ie. if you try `-m 3,10` the map will be damaged because there seem to be special patching for mission 10. The same map and level file moved one index up or down will load properly.

So in case something behaves strange, one thing you should try is to switch the mission index.
(though I will try to weed out such hard-coded fixes)

**Comment by Moburma** [comment]
That's a good shout. I tried moving the objectives to a new slot (it's mission 67) but that didn't work. It's something to do with the GAME_OBJ_GET_ITEM objective. These seem right at first glance as they point to the correct objects, but as you pick them up they do odd things and the marker gets moved to surrounding buildings. The first item objective has a coordinate set (although that doesn't work right either), so I don't know if this would be a way around the issue.
The same problem occurs if the BEN .miss file is used to play mission 65. In this version it had an objective to pick up an Uzi, but this similarly goes mental and shifts the objective to part of the scenery once collected. 
Both are original game bugs, possibly just how the objectives are set out here. Do you know the second number of GAME_OBJ_GET_ITEM should be? The first is clearly the ThingOffset of the item, is the second a uniqueid (seems like items don't have uniqueids!)?

**Comment by mefistotelis** [comment]
> it had an objective to pick up an Uzi, but this similarly goes mental and shifts the objective to part of the scenery once collected

Ha, makes sense. If you pick up a weapon, it's no longer a stand-alone thing - the `Owner` should then be used as the target.

I might fix it later; currently I'm seeing an issue with last level, probably caused by the move to text mission list.

> the second a uniqueid (seems like items don't have uniqueids!)

Sometimes there are parameters set which are unused. That's probably because type of item was changed within the editor, so at some point it was pointing an object which had UniqueId.

**Comment by Moburma** [comment]
I did more playing around with this mission and its objects. One thing I discovered, the first of the item objectives must be nonsense. It's referencing coordinates (16128,0,7914), which if placed as a GoToLocation objective type shows this is clearly a repurposed version of the very first mission objective (where you have go to the entrance of the ambush area). I thought maybe it was some way of making the game check for the object ID in the area of the weapons, but it's clearly just junk data (and why the other items don't have coordinates). There's only one other level in the game with an object collection objective that uses coordinates and that's cut, so it's probably a sign this argument doesn't actually work/do anyhing. 

Your above explanation seems to be correct as to what happens - the player collects the items, but it then can't find where they have transfered to, and they get transposed on top of pieces of scenery (and dropped enemy items if they are killed). This definitely isn't just cosmetic (i.e. just the radar) as the objective doesn't complete (or are completing objectives part of that code?). 

I just can't see why it doesn't work. The second number is definitely the uniqueID of the item. This isn't defined in the level file, but they must be right, as the objective points to the correct object. If I switch these around it will point the scanner blip to the corresponding other item object, and if I use random numbers it doesn't work at all, so it's not blind luck that it initially identifies the item.
So clearly it's something to do with the player taking the item and it being unable to figure out where it's gone. But I don't see anything special here compared to a final game level, we're just giving it two arguments, and they both seem to be correct, but the behaviour is wrong. I've also seen this happen before with mission 65 when using the .ben objectives, but that level file is not even an alpha period one, it's a final game one, so it doesn't even look like it's a final game level format difference?

**Comment by mefistotelis** [comment]
We're not far from being able to fix all the unfinished objectives. Please provide a reproduction level, and I will look into this.

Having Unguided Campaign with objectives which do not exist in the base game would definitely make it more interesting to play.

**Comment by Moburma** [comment]
It's mission 67 aka level c026l004.dat. i.e. /m 2, 67
The mission starts with an objective to go into the entrance of an enclosed area, that then triggers the item collection objective. to debug the level I usually just use the teleport cheat to get my agents next to the items, teleport one into the entrance, and then teleport him to the items and do what I was planning to do. 
Looking at the objectives and commands, what is supposed to happen is:

1. The player goes to the first objective at the mouth of the "tunnel" **First Objective**
2. The unarmed agents waiting unarmed for the peace talks grab their guns and attack
3. The player has to run to their hideout across the map unarmed
4. Player grabs their guns from the ground (hint: give a plasma lance to two unguided and the two launchers to the other two) **Second objective**
5. Once this is completed (currently impossible) it seems the unguided hiding in the buildings should pour out and attack (not sure if this works or not, they are waiting for objective 3 to be completed (WAIT_OBJ), but the base is 0 and only goes up to 2 - does 2 count as 3 from base zero?). This should probably be changed to 1.
6. The player and new friends must hunt down the much tougher second wave of plasma lance equipped agents in cars. **Third objective**

Only up to 3 above works.

I thought you did a .BEN file version, if you run mission 65 (c040l015.dat) with that set of objectives (and only that one) it gives you an extra objective to grab an uzi compared to the final version of the level, which must then be taken to the exit. This fails in the same way, mapping randomly to some piece of scenery once collected.

**Comment by Moburma** [comment]
Hmm, in all the levels I've seen, floor items seem to have a Type of 25 (0x19 in hex). Here's my tool's output of the level:


<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Andrew/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Andrew/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Parent | Next | LinkParent | LinkChild | SubType | Type | State | Flag | LinkSame | Object | Radius | ThingOffset | X | Y | Z | Frame | StartFrame | Timer1 | StartTimer1 | WeaponType | ItemName | LastFired | Ammo | Owner | OnFace | DummyData | field_38
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 0 | 0 | -356 | 0 | 25 | 1 | 67108868 | 0 | 0 | 0 | -580 | 7224576 | 0 | 4849408 | 3563 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 0 | 0 | 0 | 393216
0 | 0 | -580 | -190 | 0 | 25 | 1 | 67108868 | 0 | 0 | 0 | -356 | 7220736 | 0 | 4579328 | 3563 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 0 | 0 | 0 | 458752
0 | 0 | -356 | -521 | 0 | 25 | 4 | 67108868 | 0 | 0 | 0 | -190 | 7135744 | 0 | 4791552 | 3563 | 1003 | 5 | 5 | 5 | Launcher | 0 | 0 | 0 | 0 | 0 | 524288
0 | 0 | -190 | -407 | 0 | 25 | 4 | 67108868 | 0 | 0 | 0 | -521 | 7131904 | 0 | 4633344 | 3563 | 1003 | 5 | 5 | 5 | Launcher | 0 | 0 | 0 | 0 | 0 | 589824
0 | 0 | -521 | -140 | 0 | 25 | 7 | 67108868 | 0 | 0 | 0 | -407 | 6900992 | 0 | 4780032 | 3563 | 1003 | 5 | 5 | 8 | Flamer | 0 | 0 | 0 | 0 | 0 | 655360
0 | 0 | -407 | -146 | 0 | 25 | 7 | 67108868 | 0 | 0 | 0 | -140 | 6897152 | 0 | 4641024 | 3563 | 1003 | 5 | 5 | 8 | Flamer | 0 | 0 | 0 | 0 | 0 | 720896
0 | 0 | -140 | -139 | 0 | 25 | 16 | 67108868 | 0 | 0 | 0 | -146 | 6897152 | 0 | 4930304 | 3563 | 1003 | 5 | 5 | 17 | Plasma Lance | 0 | 0 | 0 | 0 | 0 | 786432
0 | 0 | -146 | -141 | 0 | 25 | 16 | 67108868 | 0 | 0 | 0 | -139 | 6893312 | 0 | 4506112 | 3563 | 1003 | 5 | 5 | 17 | Plasma Lance | 0 | 0 | 0 | 0 | 0 | 851968
0 | 329 | -139 | -440 | 0 | 25 | 1 | 67108868 | 0 | 0 | 0 | -141 | 1903360 | 0 | 1867520 | 3563 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 0 | 0 | 0 | 66256896
0 | 328 | -141 | -565 | 0 | 25 | 1 | 67108868 | 0 | 0 | 0 | -440 | 1905920 | 0 | 1945600 | 3563 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 0 | 0 | 0 | 66322432
0 | 327 | -440 | -138 | 0 | 25 | 1 | 67108868 | 0 | 0 | 0 | -565 | 1905920 | 0 | 2016000 | 3563 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 0 | 0 | 0 | 66387968



</body>

</html>

Another example, c040l001.dat:


<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Andrew/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Andrew/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">



</head>

<body link="#0563C1" vlink="#954F72">



Parent | Next | LinkParent | LinkChild | SubType | Type | State | Flag | LinkSame | Object | Radius | ThingOffset | X | Y | Z | Frame | StartFrame | Timer1 | StartTimer1 | WeaponType | ItemName | LastFired | Ammo | Owner | OnFace | DummyData | field_38
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | -588 | 0 | -590 | 0 | 25 | 1 | 67108868 | 0 | 308 | 64 | -591 | 1028608 | 0 | 2159360 | 4411 | 1003 | 5 | 5 | 2 | Minigun | -2000 | 0 | 4763 | -46 | 23554 | 73859393
0 | 0 | -591 | -589 | 1 | 25 | 6 | 67108868 | 0 | 256 | 64 | -590 | 6230784 | 4.29E+09 | 5697024 | 4411 | 1003 | 5 | 5 | 0 | Briefcase | -32 | 0 | 5 | 3 | -22 | 71827698
0 | 225 | -590 | -588 | 0 | 25 | 1 | 67108868 | 0 | 256 | 64 | -589 | 975616 | 0 | 2061056 | 4411 | 1003 | 5 | 5 | 2 | Minigun | -32 | 0 | 321 | 3 | -22 | 73793778
-591 | -584 | -589 | -587 | 0 | 25 | 1 | 67108868 | 0 | 308 | 64 | -588 | 1028608 | 0 | 2159360 | 4411 | 1003 | 5 | 5 | 2 | Minigun | -2000 | 0 | 4763 | -46 | 23554 | 73859393
0 | 220 | -588 | -586 | 0 | 25 | 1 | 67108868 | 0 | 256 | 64 | -587 | 1001216 | 0 | 2745088 | 4411 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | -11 | 0 | -12 | 74055919
0 | 221 | -587 | -585 | 0 | 25 | 1 | 67108868 | 0 | 256 | 64 | -586 | 1023488 | 0 | 2906880 | 4411 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | -17 | 0 | -1 | 74121455
0 | 224 | -586 | -584 | 0 | 25 | 1 | 67108868 | 0 | 256 | 64 | -585 | 1008896 | 0 | 2276864 | 4411 | 1003 | 5 | 5 | 2 | Minigun | 0 | 0 | 13 | 0 | -10 | 74252288
-588 | 226 | -585 | -583 | 0 | 25 | 1 | 67108868 | 0 | 308 | 64 | -584 | 1028608 | 0 | 2159360 | 4411 | 1003 | 5 | 5 | 2 | Minigun | -2000 | 0 | 4763 | -46 | 23554 | 73859393
0 | 217 | -584 | -582 | 1 | 25 | 30 | 67108868 | 0 | 0 | 0 | -583 | 2686464 | 0 | 2999040 | 4616 | 1056 | 5 | 5 | 31 | Epidermis 4 - Stealth Skin | 0 | 4 | 0 | 0 | 0 | 74579968



</body>

</html>

I thought they were always Type 25, and then Subtype 0 for weapons, 1 for generic items (generic briefcases and epidemises) and 2 for money briefcases, unless I'm missing something.

**Comment by mefistotelis** [comment]
Yes, you're right.

So it looks like two of the three TNG_CMD_GET_ITEM commands were working for similar reasons as boarding the mech - the thing index was incorrect, but when the game loaded the level completely, it matched.

I guess after my fix, no Agent will be able to pick up the gun. For that to work, the level needs to be updated.

Currently what the agents are trying to pick up are:
```
Sync: SWars: fix_thing_command_indexes: Thing-141 at (1620,0,4941) type=5,2
Error: SWars: fix_thing_command_indexes: Cmd46 = TNG_CMD_GET_ITEM target person/vehicle/item -141 not found
(points to a static light rather that dropped item)

Sync: SWars: fix_thing_command_indexes: Thing-440 at (29221,0,22316) type=5,2
Error: SWars: fix_thing_command_indexes: Cmd52 = TNG_CMD_GET_ITEM target person/vehicle/item -440 not found
(points to a static light rather that dropped item)

Sync: SWars: fix_thing_command_indexes: Thing-565 at (0,0,0) type=0,0
Error: SWars: fix_thing_command_indexes: Cmd53 = TNG_CMD_GET_ITEM target person/vehicle/item -565 not found
(points to a non-existing thing)
```

**Comment by mefistotelis** [comment]
Is `GAME_OBJ_PKILL_G` used anywhere? Does it work?

Because the implementation is strange. It checks only first (the last added) member of the group.
Shouldn't that check if at least certain amount was killed?

**Comment by Moburma** [comment]
Checking today after your latest commit, the items bug here is fixed! Level seems to work great now, the objective completes and the other unguided DO all come out once the player has all the guns.

As for GAME_OBJ_PKILL_G, it is used in the .ben file. I need to dig out my exports, but I'm pretty sure it was originally used in the demo level there. It changes after that, so I can only assume it did not work very well as it never appears again afterwards.

**Comment by mefistotelis** [comment]
> As for GAME_OBJ_PKILL_G

Yeah, I don't see how the current implementation would be useful. "Kill one group member, but not any member, it must be the one which was added there last" doesn't seem useful to me. There is a separate objective for killing specific person.

I will fix that command into `GAME_OBJ_PKILL_MEM_G` - kill given amount of group members.

**Comment by mefistotelis** [comment]
Idea: Maybe Punks should be able to hack/persuade spiderbots or deployed IFFs?

**Comment by mefistotelis** [comment]
Oh, I see the problem - we've loaded mission INI file, but then we did fixups to it. So with the fixed data, we cannot expect that it will survive re-applying fixups on an already "fixed" set of data.

For the hacking - maybe it would match more as a scripted event rather than generic game mechanic.
Ie. in Unguided version of the mission where you kill a hacker, a spiderbot changes its group assignment and start attacking Zealots.

**Comment by Moburma** [comment]
That's a much better idea! I think that is another angle that would be interesting and is already present in the final game, the idea that because the Unguided are fundamentally weak (they're just normal people, not cybernetically enhanced killing machines) they use "tricky" tactics like this. Would be a fun mission objective to turn the tables on enemies this way and fits in with psycho gas as well.

**Comment by mefistotelis** [comment]
> So it looks like two of the three TNG_CMD_GET_ITEM commands were working for similar reasons as boarding the mech - the thing index was incorrect, but when the game loaded the level completely, it matched.

This turned out to not be true - the index was correct! There was just a sneaky bug in the code.
Now all the 3 guns are piced up correctly  (in `-m 2,67`).

**Comment by Moburma** [comment]
Awesome! I actually made a video of the level last night :
https://youtu.be/Hcn3_VQFiHc
It's a really fun level and works perfectly now. Brutally hard (you do not want to know how many attempts this took to beat...), but then it is meant to be late game. 
Also makes me think how useful the Chromotap would be for Unguided levels, would be a good reward/starting kit on late levels to make up for their pitiful health and no mods. Another example of an existing but undervalued game item that could play a major role in the campaign.

**Comment by mefistotelis** [comment]
>  I can't see anything that does that in the .miss file, so I assume this must be hardcoded?

No, it's not in the code. Who you control depends on the default group. If there are Agents/Zealots in the group, they are replaced by your team. If not - you control whatever was in there.

I did a modification there: if there's no person in the group, the game tries to fix player group index by finding a better one. If even that fails, it duct tapes an agent in there. This way, the game does not crash due to no group - this fixed several unfinished missions.

If there is more than 4 people in the players group, the rest is probably quasi-deleted.

**Comment by Moburma** [comment]
Ahh of course that makes sense. I saw the adding an agent addition, that was helpful. It's interesing as this WAS already a feature of the beta version, if a level has no player units in that build it puts in 4 Syndicate agents at a fixed place in the level. They must have taken that out for the final game, in a similar way to how the simulated levels degraded for the final (probably as both were debug functionality they didn't want players messing with).

**Comment by mefistotelis** [comment]
De-persuasion idea:

Your mission is to get into a compound, separated by gate. No empty cars around, and no explosives. The mission inside can be either assassinate or get item.

Zealots are persuading people in the area. They have a car, but there's always someone inside - you cannot steal it. Your mission is to allow one of your punks to get persuaded, and this way he can be transported (with the rest of the crowd) via car to a separate Zealot compound. After he is there, you must somehow help him de-persuade (maybe there's a fellow punk with Disruptor who will help him?), then do the mission, and escape.

**Comment by Moburma** [comment]
I think that's a neat idea, do you know if the game can even handle your units being persuaded? I think de-persuasion and the disruptor are also really overlooked mechanics in the game, could easily create a level around getting people back who have been persuaded by Zealots, especially if the targets are weak and prone to dying in combat, dissuading just blasting the zealot.

**Comment by mefistotelis** [comment]
> do you know if the game can even handle your units being persuaded?

I don't see why it wouldn't. Haven't tried though.
You definitely need more agents, as if all are persuaded that's game over.

**Comment by Moburma** [comment]
Something I have been working on:
[<img src="https://i.ibb.co/6b6WWxZ/swleveled.png">](https://ibb.co/6b6WWxZ)

Early days yet, doesn't do much useful, but it will. I didn't want to do it in bloody PowerShell again, but I got bogged down learning C# and wanted to make some progress so started in it anyway. Someone is helping me move my Magic Carpet editor to C#, so I will use that as a template and eventually move this to it as well.

**Comment by mefistotelis** [comment]
Looks nice.

I looked into objectives which are build into levels - looks like they're outated version of the ones from missions.
But just for reference, here are all of them. Only few levels even have those:

```
-m 0,4
P00 = GAME_OBJ_PERSUADE_ALL_G( Group(1), Arg2(1) )
-m 0,23
P00 = GAME_OBJ_PERSUADE_P( Thing(743,1), Arg2(1) )
P00 = GAME_OBJ_P_ARRIVES( Thing(749,7), Coord(15171,0,27653), Radius(19) )
-m 0,59
P00 = GAME_OBJ_MEM_G_ARRIVES( Group(0), Coord(30649,0,29872), Radius(7), Amount(1) )
-m 0,88
P00 = GAME_OBJ_GET_ITEM( Thing(-98,1083) )
-m 1,40
P00 = GAME_OBJ_GET_ITEM( Thing(-740,1109) )
P00 = GAME_OBJ_PERSUADE_P( Thing(18,1137), Coord(5862,0,31284), Radius(24) )
-m 1,44
P00 = GAME_OBJ_ALL_G_ARRIVES( Group(0), Coord(15241,64,7036), Radius(13) )
P00 = GAME_OBJ_DESTROY_OBJECT( Thing(43,0), Coord(17996,7,953) )
-m 2,67
P00 = GAME_OBJ_NONE(  )
P00 = GAME_OBJ_NONE(  )
P00 = GAME_OBJ_GET_ITEM( Thing(-459,12), Arg2(17) )
P00 = GAME_OBJ_GET_ITEM( Thing(-456,9), Arg2(5) )
-m 2,78
P00 = GAME_OBJ_DESTROY_OBJECT( Thing(583,0), Coord(20117,0,17561) )
P00 = GAME_OBJ_MEM_G_ARRIVES( Group(0), Coord(8021,384,28268), Radius(10), Amount(1) )
-m 2,82
P00 = GAME_OBJ_DESTROY_V( Thing(116,1004) )
P00 = GAME_OBJ_DESTROY_V( Thing(116,1004) )
P00 = GAME_OBJ_DESTROY_V( Thing(116,1004) )
P00 = GAME_OBJ_DESTROY_V( Thing(116,1004) )
-m 3,17
P00 = GAME_OBJ_PERSUADE_P( Thing(296,1), Arg2(1) )
-m 3,42
P00 = GAME_OBJ_GET_ITEM( Thing(0,1000) )
P00 = GAME_OBJ_GET_ITEM( Thing(0,1000) )
P00 = GAME_OBJ_GET_ITEM( Thing(0,1000) )
```

**Comment by Moburma** [comment]
Ahh, that's very interesting, I never noticed them before. I always thought it was a bit of a strange idea to take the objectives out of the level file in the first place... As far as I can tell the PS1 version DOES (or at least _can_) use objectives inside the levels files, it only seems to use .mis entries for the initial objective text when you start a level, but it correctly tells you what to do if you rename level files to play them.

On the topic of level structure, I have a couple of things I wonder if you know about:

- Some levels have some kind of header or data after the command data and before the group names. Usually it's 01020304050607 but there a few variations. Have you seen anything referencing this? Not sure if it is used or not and if so what it means
- Some levels have what looks like 4 bytes of data at the very end of the file - but many others don't. Don't know what this means yet
- There doesn't seem much rhyme or reason to how much zero byte space there should be at the end of a level file yet, but I haven't looked into this much yet.

**Comment by mefistotelis** [comment]
>  I always thought it was a bit of a strange idea to take the objectives out of the level file

There clearly was an idea with some of your changes to the level being transferred to when you re-visit the city. Though the idea wasn't well thought through.

> Some levels have some kind of header or data after the command data and before the group names

There's 1320 bytes of data there; not sure what it does, but it's stored in a similar manner to level names - I guess it's a scratch space for unimportant stuff.

> Some levels have what looks like 4 bytes of data at the very end of the file - but many others don't.

Aha.

```
    if (fmtver >= 16)
        LbFileRead(lev_fh, &dword_176D58, 4);
```

I didn't gave it any name proper name, so I didn't figured out what it does.

> There doesn't seem much rhyme or reason to how much zero byte space there should be at the end of a level

```
    if (fmtver >= 10)
        LbFileRead(lev_fh, game_level_miscs, 4400);
```

Just looked into this one. It's some kind of array which can be set to influence buildings on the map. Actually only one `Type` of the field is implemented, and that type (=1) seem to search for  Mounted Guns on the map and select which weapon is mounted on top. So apparently, mounted guns can shoot with any weapon.

**Comment by Moburma** [comment]
Another question, do you know how/where the scanner minimaps come from? Are they stored in the map files (or elsewhere), or are they generated by the game based on the actual map? The reason why I ask is that of course there are not .SCN files for all the levels, so for example there is no scanner map for Anchorage, which would be needed for the Tale of Two Cities level in the Unguided campaign.

**Comment by mefistotelis** [comment]
If I remember right, menu uses SCN files while within the 3D engine the game generates a scanner bitmap at start, from the actual level data.

The generation is quite computation costly, so it is done only at start of the level, and rectangular updates are done after a building gets destroyed.

To get the missing SCN, you'd have to just write it to file after it's generated.

**Comment by Moburma** [comment]
I've got the file handler working and outputting data, but... I don't understand _where_ to get the data from!  I can't see SCANNER_data defined anywhere and I can't figure out which of the scanner functions draws the map (SCANNER_init?). Sorry if it's obvious, but I can't work it out, could you hand-hold a little more if you have time?

**Comment by mefistotelis** [comment]
> I don't understand where to get the data from! I can't see SCANNER_data defined anywhere

Oh right, that makes it a little more complicated.

1. You need to export the SCANNER_data symbol from assembly - in `swars.sx`, replace all uses of `SCANNER_data` with `EXPORT_SYMBOL(SCANNER_data)`. And the last use will be the actual data definition - `SCANNER_data:`. This one, you  need to replace with `GLOBAL (SCANNER_data)`. Note that the `:` needs to disappear.

After that is done, the code should compile like normal. But now you can define that variable in C by using:

```
extern ubyte SCANNER_data[256][256];
```

Why all that is needed - in C, symbols are "decorated" meaning some prefix and/or postfix is added to them. In assembly, you work on raw symbols - no decoration is applied. So if you have some data defined in ASM, and you want to see this data in C, you need to decorate the symbols in the same way C compiler does that.

**Comment by Moburma** [comment]
That did it! I had a few compile errors so I had to revert some references to SCANNER_data, but it worked!

<img src="https://i.ibb.co/nfhZVHM/MAP060.png" alt="MAP060" border="0">
<img src="https://i.ibb.co/mFVNq0z/Desktop-Screenshot-2023-09-02-18-53-32-45.png" alt="Desktop-Screenshot-2023-09-02-18-53-32-45" border="0">

I've now got .SCN files for Anchorage and the Moon. Do you want them in the Levels repository?
I also got one for the early version of Anchorage, but I get the impressioin that's not really supposed to have been included with the game.

**Comment by mefistotelis** [comment]
>  now got .SCN files [...] Do you want them in the Levels repository?

No; I think we'll have the .SCN files auto-generated at some point, so no need to have them within the repo.
(unless the generation won't be ready on time, then we'll add them temporarily)

**Comment by Moburma** [comment]
A few things:
I see you put basic objective text for the unguided levels, what I was planning to do was write up some basic step by step objectives for each level. This would just be matter of fact, no attempts at story or charaterisation, just a matter of fact list of "1) kill the Agents 2) get the briefcase" etc. The idea this would be so people could playtest levels and know what they are doing (something essential for e.g. the "Rescue" level that makes no sense without a briefing). I might start on this tomorrow, it won't take long

I reached out to Sean Masterson who wrote the story for Syndicate Wars to see if he remembers about the game and especially the Unguided campaign. I don't know if I'll get an answer, but you can but try with these things. If he does hopefully I can get some mysteries solved.

Did you mean to put the new Unguided level names in the main NAMES.TXT? Sorry if I missed something, but it looks like some of the original level names are replaced as a result (e.g. the first eurocorp mission Executive Alert). Also we should use the demo version of this file that includes The Shining. I can do a pull request for it.

**Comment by mefistotelis** [comment]
> I was planning to do was write up some basic step by step objectives for each level. This would just be matter of fact, no attempts at story or charaterisation

Yes, I was thinking about the same - we first need a "tactical level" description, then we may (if we'll get there) add a story (aka "strategic level") to it.
The tactical descriptions would not only allow to play the levels right now, but also would help in creating the story.

> Did you mean to put the new Unguided level names in the main NAMES.TXT? Sorry if I missed something, but it looks like some of the original level names are replaced as a result

What I want to do is create separate WAD/IDX for each campaign. As it is right now, there are some issues.
But I did forgot about level names; this can be fixed by commenting out the problematic strings from Unguided file.
(the other issues are with weapon descriptions in non-english translations, due to translated weapon name being used to find the description)

**Comment by Moburma** [comment]
Prepare to have your mind blown. From Mr Masterson himself:

"I remember the outline with which I am sure you are already familiar: the Unguided are an unforeseen consequence of Eurocorp losing its grip on elements of the populace at large (thanks to the efforts of the Church of the New Epoch, which has also failed to foresee the rise of the punks). However, given the sneakiness of characters like the double agent, Lucy de Saxo, I am certain that we had some plot-twisty plans for the Unguided. What exactly they might have been is almost certainly lost to time now.

I also remember another thing. Are you familiar with the Public Access Network elements? These were never properly developed due to time constraints but they were originally intended to be Easter-egg-like sources of bonus mission goals. And I believe we intended the mechanic to feature heavily in the Unguided campaign.

As with any game development, we sometimes had to abandon ideas that were too grandiose, due to time, budget or technology. So, while there was talk of the Unguided campaign being unlocked after completing the central storyline, we entertained other possibilities. Imagine choosing to play as Church or Syndicate but, whatever your first mission, if you accessed a PAN node (by accidentally causing an Unguided character to pass too close to one), you would open up an alternative, Unguided-focused mission objective and, consequently,  the rest of the Unguided campaign. Because, let’s face it, Church or Syndicate, whichever side wins, it’s not going to spell Happy Ever After for the serfs. But developing that? Technical challenges aside, we would never have had the time or budget. "

He said he wouldn't mind answering other questions, so let me know if you have anything particular to ask him. I've got loads of things already to ask, but want to make the most of his kindly offering to dredge his memories.

**Comment by mefistotelis** [comment]
Yeah, I was thinking about the plot twist. There should be one. There are many missions with prof. Drennan, so maybe he could be used for that - but he does die, finally. I didn't come with anything very interesting.

For PAN - well I did had a different view at it. I was imagining something like the game "Uplink", where you hack servers and play minigames to do so. Unlocking the Unguded the Masterson way suggests that the player is a double agent as well. That is an interesting twist by itself.

I'd ask about any old disks he has - for DK we were able to get some, and while it didn't helped with reversing the game, it was interesting to see various experiments - ie. tries to make creatures switch weapon and armor (the Armory building was intended as storage for such items, and creatures would go there to get dressed for combat). If the disks have bad sectors, it doesn't matter - we can restore the content, reading raw sectors.

For the Punks - I wonder if the player was supposed to choose, like in "Original War", whether to go with original agenda or switch to Punks for few last missions. Because if there is no clear branch point, it's also possible the player would just get alternate objectives - ie. you get an email with assassinate mission, plus additional Unguided email about persuading the target and escorting to a building instead.

**Comment by Moburma** [comment]
Yeah, I edited that bit out, unfortunately he already said he doesn't have any development materials from those days. 
PAN I always thought it would just work essentially like a switch in-game to open doors or something, and to unlock some kind of extra lore info in the menus, but it seems I was totally wrong. This game is quite the rabbit-hole.

**Comment by mefistotelis** [comment]
>  I guess there could be some way to build up to it to write it that the player could genuinely believe that Eurocorp would surrender

In Eurocorp campaign we do get rid of parts of our Syndicate which have rebelled, or even not rebelled but they might do that, so we clear them just in case ("we will replenish the human resources when our cloning facilities are back on-line" or something along that).

**Comment by mefistotelis** [comment]
For some reason, in Unguided campaign, city names are shifted by -1 (ie. Detroit(1) becomes London(0)).
Wierd.

**Comment by Moburma** [comment]
Ahh, I noticed that last week, I assumed you just weren't finisheed working on it, I didn't realise this was actually how the game was. Guess this probably dated back to pre-alpha times.

**Comment by mefistotelis** [comment]
> I assumed you just weren't finisheed working on it

Well that's true.
Regardless, I figured it out - the issue is with translation text files, not with list of cities. The text file starts with names of city properties, but there are only 2 sets of these, so the 3rd campaign consumes the first city as the names of properties.

**Comment by mefistotelis** [comment]
With the latest commit, the campaign no longer stops at the city of Anchorage with older map - it now considers both 52 and 60 to be the city of Anchorage.
So the campaign should now be passable, from the missions flow standpoint.

**Comment by Moburma** [comment]
That's cool, although actually I'm not sure it's needed - I think it's best to consider map 52 as unused. No actual levels use it, C052L001.DAT and C052L007.DAT are completely unplayable as - is as the Thing placement cannot ever match the terrain, therefore they need to be transplanted to a new map to work (map 79 was the only I found that made sense). Having said that, I could see in the future maybe something like radically different versions of the same map with say large parts of the terrain destroyed that the player returns to that this would be great for.

**Comment by Moburma** [comment]
I have updated the Unguided campaign to now have mission objectives for all levels (such as where they were never set by Bullfrog) and simplistic briefings that just list what must be done, as well as netscan entries to draw attention to points of interest. 
This process also revealed a few things about some levels:
Mission 80 (aka Clean up) is a weird one. I realised there is actually a third group of protesters walking around in circles towards the left hand edge of the map, but this just seems to be a leftover from experimenting rather than having any meaning. Still not sure what to do with this level, it could be good as maybe some kind of Tower-defencey style thing where you have to use Cerberus IFFs to stop Agents on foot attacking, but at the moment they just fly in and either have to be killed instantly or they just massacre the civilians making it impossibly hard and quite simplistic anyway.
Also Mission 60 (early version of Guns Guns Guns) doesn't really work. Looking at it more closely, it's obvious that the map must have been different at some point, and the soldier at the bottom wall of the base must have been standing at some kind of door or other building. It makes no sense for him to be stuck in a wall, and also exposes another problem - I set mission objectives to get the items there, as that is clearly the intention of the level (going there triggers the counter-attack from agents). However, as you have to currently destroy the wall to get in, this also destroys the items. This in turn breaks the objectives to pick up the now gone items! The deeper I got into it, the more I realised it was a mess. I quite like the level, so I guess what needs to happen is to move the soldier into the base. I do wonder if he is actually even meant to be in the base, as there is usually an agent trying to shoot him through the wall, so perhaps they were supposed to an unaffiliated group who you would get weapons from first, then attack the base? Something changed at some point though, and it just makes no sense now.
I also see the second group of punks have logic to do something, I need to work out why that keeps failing.

I also need to dig out my old edited version of the newer Communications level that I hacked up to work on the Rekyavic map. I moved the grenade item down so it could be collected, but the item object bug kicked in so the objective never completed. That should work now, but I guess the destroy building objective will likely be wrong as it's clearly for some other map. Therefore some building would have to be chosen as the new target. Might be salvagable, though.

I also did a lot of research into the characters and plot I need to write up here, such as Unguided characters mention in the other campaigns.

**Comment by Moburma** [comment]
Actually, Communications works pretty good now on the Reykjavik map. With the item bug fixed, it actually choses an entirely appropriate building as the bombing target on the explovies are collected - a bunker in the middle of the level. If I didn't know better I would swear it was always intended to run on this level. There are still lots of enemies stuck in walls, but it plays pretty good. 
Also made me realise a) the trees on that map are the same ones from the early Magic Carpet build (actually from Populous I believe) and b) that the fact you can set trees on fire and they kill people is also an idea from Magic Carpet.

**Comment by mefistotelis** [comment]
you're right.
Weapons can already be given as "researched". See how Persuader II is given to the player in Eurocorp campaign.

**Comment by Moburma** [comment]
Spent a lot of time looking at a few levels over the bank holiday here. 

Some discoveries/comments:

**Mission 70 Guns Guns Guns**
I still don't get wtf the point of this one was going to be, I will just have to assume it is very unfinished. Clearly you're supposed to enter into the small base at the top to kill the scientist, but achieving that seems impossibly hard, as in I can't see this being possible unless you had four high level punks with you. 
An important thing I discovered is that actually the zealots in the top base are set to board the vehicles there and leave the base undefended if any of the zealots in the very middle of the huge base are killed. This opens some interesting gameplay possibilities, but is still contradictory. If you're strong enough to get into the middle base in the first place then you're not going to care about getting rid of the zealots at the top! I tried a different tact to work around this - giving the punk a clone shield. This worked quite well, there's just enough juice on him to sneak into the middle of the base. Killing any of the zealots there is difficult but doable, but there really need to be fewer for this to work. It's then possible to sneak back out, past the approaching new zealots and kill the scientist in the undefended base (well um, apart from the ambush that happens when you kill him...). Another problem here - the zealots in the flying cars weirdly got stuck going over the gate when I actually did this. Still not decided what to do with this one.

**Mission 60 Guns Guns Guns**
The older version of this. I really sat down and think I properly understand what was intended for this one now. I always liked this as the whole round up punks and attack en masse concept is really fun. Now it was always clear you were supposed to get the punks to help you, but I couldn't get the second group to move. Checking them in the debug mode, actually they were waiting for YOU to get into one of the ambulances. Doing this causes them to group up and attack the base. Clearly there must have been an objective to get in the car originally for this to make sense, so rounding up the punks likely was not optional but instead a core objective with actual markers for the places to go. 

Then there's the whole thing around the man stuck in the wall. This clearly shows that originally the building must have had a different design, with at least a room there and possibly a back entrance to the heart of the base. He gives a clue what the objective probably was - he has a nuclear grenade and is tasked with blowing up one of the towers in the main area (which he can't do due to being stuck). Along with the fact his stockpile has 3 grenades it's likely the main objective was to blow up these towers. This also triggers most of the eurocorp agents along the way. Finally there are a few Eurocorp agents in the top base that don't do anything - the obvious inference being to kill them too. There is also a bike in there though, so maybe it was to get that back (and thus causing them to attack while doing it)? 
I edited the level to fix it up, moving the soldier and his stockpile into one of the corner bunkers (god knows how he is meant to have got in there, but that's a problem with the original level design), so now all that part works (well, his pathfinding keeps making him try to walk through the wall where I think there was a door, but I just need to add more go to commands to work around this). 

I'm quite happy with it so far, so will likely release it soon. One thing is has revealed though is that I don't think the vehicle related objectives (GAME_OBJ_ALL_G_USE_V and GAME_OBJ_MEM_G_USE_V) actually work (or it could be me using the wrongly). @mefistotelis  Have you ever tried these? I was using them like this GAME_OBJ_ALL_G_USE_V( Thing(197), Group(0)) but the game never updates to them this way. I wonder if they are bugged -  they're not used in any levels that I can see, and it would explain why e.g. the bullion truck levels just used suitcases for the money in the end.

**Comment by mefistotelis** [comment]
Looking at definition inside the code:
```
{"GAME_OBJ_ALL_G_USE_V", "ALL USE VEHICLE",	ObDF_ReqVehicle|ObDF_ReqSecGrp|ObDF_IsAlly },
```

So the use should be `GAME_OBJ_ALL_G_USE_V( Thing(197,<uniqueID>), SecGroup(0))`.

For:
```
{"GAME_OBJ_MEM_G_USE_V", "MEM USE VEHICLE",	ObDF_ReqVehicle|ObDF_ReqSecGrp|ObDF_ReqAmountY|ObDF_IsAlly },
```
it only differs with the amount stored in "Y"... but I don't think such storage of amount is supported in code. So that's probably broken - I mean it is implemented, but the INI parser can't do what is necessary.. I guess using Coord(0,<amount>,0) could workaround this?

In general `Thing()` and `Group()` will not work in one command, if two are required (2x thing, 2x group, or one of each), one definitely should use second storage (be `Sec`). For which one - that needs to be checked in the code.

I should probably improve the parser to log an error if params are clearly invalid.. and fix the Amount.

**Comment by Moburma** [comment]
**Mission 69 Trash Church**

Had another look at this one to perhaps restore it, as there are loads of commands for deleted Things. Interestingly the ones that exist correspond with the correct MyThing value, so we can accurately see what was there and what these NPCs did (they're not just duplicates of existing numbers as seems to have happened with later versions of the game's editor). By my count there would be at least 62 more zealots in the level. The problem is there's no position data (i.e. go to point commands etc) to hang placements around, and even more pertinently - it's so boring. 
It was a very very simplistic level, it must have been developed very early on as you really have to go back a ways to find other levels as simplistic command wise (more sophisticated commands likely didn't exist yet). It really is at the level of the original Syndicate, with nearly all of the enemies set to just stand still until the player kills all the first group of zealots and then just attacking en masse. 

The one interesting thing is there are characters setup to attack if the player kills members of the civilians group, not sure what that would be about.

There are also a couple of NPCs using WAIT_TIME set to 9999, which as far as I know means indefinitely. But weirdly they then both have LOOP_COM commands to repeat them, so maybe that didn't work back then?

**Comment by mefistotelis** [comment]
>  there are loads of commands for deleted Things

In the past, I often used some programs which had bugs. One of these was CorelDRAW - innovative for its time, allowed to create complex images. Working on it' you'd quickly learn to save every 15 minutes, to make sure you won't be set back too much when in inevitably crashes or just stops working in some way. But even then, sometimes you'd still forget that and work for 2 hours straight, then see a "save error".

A method to not lose the 2h progress was then to start deleting things, until it saves - because usually there was just one broken object somewhere (like Bezier curve around a single node, or non-closed shape with properties of a closed one).

I wonder if this was the same - someone started deleting things in hope at some point it would start working.

**Comment by Moburma** [comment]
> I wonder if this was the same - someone started deleting things in hope at some point it would start working.

Maybe, my personal hunch was someone deliberately starting again. Looking at it it was saved after the last version of "Talks Ambush" which is an absolute masterwork of what was possible with the early game commands. In comparison this one is "my first level". The stuff that was left was mostly things that were boring to set up like people driving around (although there seem to have been lots of the usual wandering civilians not kept) that would be useful as a template. They didn't get far and I guess were told to stop working on Unguided levels not long after.

**Comment by Moburma** [comment]
I've been doing quite a lot of work on this since I have my level editor tooling at a standard I can work with. Main focus has been expanding the bare bones of the barely started "Trash Church" and make it into an actual proper level. This is coming on slowly as I've learned how things actually work and behave and what's possible with the game. I've got a fair bit done now, but a lot of it is scripted sequences and little vignettes, so I need to get more meat of the level done. 

I decided on:

- As the level name suggests, a basic kill everything/one level
- Late game difficulty/tech
- Gathering other unguided to help you 
- Tech tree in Microcosm - that is you start with slightly weak ordnance, but some of the enemies have very powerful weapons that you will need. Of course, you will have to take them off them, first...

Some pictures:

![](https://i.postimg.cc/TdpTRnNS/Desktop-Screenshot-2024-06-16-22-57-27-22.png)
![](https://i.postimg.cc/qJS4gg97/Desktop-Screenshot-2024-06-16-22-57-52-53.png)
![](https://i.postimg.cc/4GSJWFmT/Desktop-Screenshot-2024-06-16-22-58-11-30.png)
![](https://i.postimg.cc/hgQK6bdQ/Desktop-Screenshot-2024-06-16-22-58-48-92.png)
![](https://i.postimg.cc/rFynkNQR/Desktop-Screenshot-2024-06-16-23-02-47-76.png)

**Comment by mefistotelis** [comment]
Sounds good.
I suspect we will have more Unguided levels which require the player to arm the team first.

**Comment by Moburma** [comment]
I made a discussion topic for the above level here:
https://github.com/swfans/swars/discussions/181
Go there to play an early version of it.

This topic should probably be a discussion topic rather than issue as well...

**Comment by Moburma** [comment]
> Just looked into this one. It's some kind of array which can be set to influence buildings on the map. Actually only one `Type` of the field is implemented, and that type (=1) seem to search for Mounted Guns on the map and select which weapon is mounted on top. So apparently, mounted guns can shoot with any weapon.

Going back to to this, did you ever make any other notes on it? I've been looking into this today, mostly got it worked out, but a couple of odd things. How it seems to work is this:

6 byte header, I don't know what the first 2 bytes mean, as they are different per level file and I can't see what they correspond to. Then the 3rd byte is always 1, then the last 3 bytes are blank. Then you have 110 bytes (presumably 5 x 22 bytes) of either padding or some other usage (I've seen a couple of levels with one or two seemingly random 16 bit values in here, but usually it's blank). Then you have the first entry for a mounted gun (if there actually is one on the level). 

The actual gun entries are 22 bytes long as documented in the code.  A rough sketch of the structure is something like this:

short X;
short unknown3; [y?]
short Z;
short group;  
short unknown9;
ushort unknown11;

and the rest is just padding

I'm absolutely certain X and Z are position data, but don't know what the second parameter does (seems to be low numbers). The next bit seems to set the gun to be aligned with that group, and will then engage enemies of that group (this also updates in real time like it was an NPC, so if a neutral group becomes that group's enemy, it will then shoot on them too). The other two bytes I don't know what they do, unknown9 I can't work out at all what it does, but unknown11 does some weird things if set above a value of 1, seems to put the gun on the player's side and will engage enemies only if they shoot your units? 

Anyway, don't worry if you never got further than that, I'm quite far into figuring it out, just wondered if I was wasting my time. 

I also spent ages trying to get the gun in the dome on the London map working.... Then realised it's not actually a real gun, is it? There's no mounted gun object in the map file, so I think it's just a static 3D object and not a working gun. 

Another weird thing is there are 5 guns setup in the misc section of the Bangkok level The Dead Zone when there are just 3 on the map. I think this is just a mistake and from testing it seems the later entries take precedence over the earlier ones.

**Comment by mefistotelis** [comment]
> did you ever make any other notes on it?

I can't remember anything. But just getting back to it now, this is what I have:

```
struct LevelMisc { // sizeof=22
  short X;
  short Y;
  short Z;
  short Group;
  short Weapon;
  ubyte Flag;
  ubyte Type;
  ubyte field_C[9];
  ubyte field_15;
};
```

Should probably add it to the port.. anyway, it would seem I didn't assumed there's any header, just 200 entries of size 22 each.
The memsystem entry seem to confirm that:
```
MemSystem mem_game[] = {
[...]
  { "level_miscs",		(void *)&game_level_miscs,		22u, 200, 0, 0, 0 },
```

**Comment by Moburma** [comment]
Thanks, that confirms what I found. Sadly changing Weapon doesn't seem to do anything, it always uses lasers no matter what the value is (flamer or plasma lance would be interesting...). Flag definitely works, but I don't fully understand what it does yet, as above it seems to change something around group/enemies. 

> it would seem I didn't assumed there's any header, just 200 entries of size 22 each.
> The memsystem entry seem to confirm that:

I'm pretty sure there is. It's most obvious on levels with no items and no guns, then you get this floating around (from c030l001.d1):

![](https://i.ibb.co/LgkzM3n/c030l001d1.png)

From my experiments of getting the starting camera angle working ages ago, if this header isn't in place in the file it won't read the rest of the file properly (i.e. the file can be the correct length, but it will just ignore any values set as it didn't find what it was looking for).

I need to check if this comes BEFORE the start of the misc section, or in it. Too hung over at the moment! It's definitely right next to the last item entry, if there are items in a level.

It seems to be 3 shorts, first one always has a 16 bit value I don't understand the significance of, then it has the second one which is always 1, then it has an empty 00. Then the gun data starts. It always has 5 empty entries of 22 bytes before the guns begin, but rarely there is some other data in these 5 entries, so I'm not sure if this first bit is reserved for something special, or this is just leftover junk data.

Example is c003l007.d1, but this doesn't have any guns data. I've seen these show up in ones with guns too, and again the guns don't start until the fifth entry (110 bytes in from the header). You can also see the header again here.

![](https://i.ibb.co/2Kdqt34/c003l007d1.png)

> Modifying `engn_*c` changes the starting coordinates of player camera - so it looks like if there's any such gun on the level, the game starts by traveling from said gun to players position. A bit wierd.. but I guess why not. Or maybe these values are overwritten later, I'm not sure.

Yep, it definitely does, I had this when playing about with it. Weird design choice. I put in bad data once and it started off in the void.

**Comment by mefistotelis** [comment]
>  changing Weapon doesn't seem to do anything, it always uses lasers no matter what the value is

Right, looking at the implementation, the value is ignored. I'm not sure it was always like this though, maybe it was a workaround for a bug.

I just tried level `0,29` with some extra logs, and it tries to set weapon 60. If this wasn't ignored, it would probably crash the game - there are only 32 weapon slots.

```
Sync: SWars: level_misc_update_mgun: Mounted gun at (19082,1653) set to (60)
Sync: SWars: level_misc_update_mgun: Mounted gun at (19085,7808) set to (60)
```

**Comment by mefistotelis** [comment]
>  there are two clear other large things defined

These look like just remains of previous entries. They're not even of correct size - they have 20 bytes, the entry now has 22.

Though if we have some old levels where there are only 20-byte entries, we could fix out old levels support to read that.

**Comment by Moburma** [comment]
Yeah, I assumed it was either old junk data or maybe something else in the level. I doubt there's much interesting, it's probably just an old way of doing the mounted guns that got replaced. I'm not sure what format it is in (assuming it's either shorts with the less precise location data format doesn't make sense, and neither does 3 longs of the more precise sort), but I don't think it's worth looking into more than that.

**Comment by mefistotelis** [comment]
I checked all the levels, and we do not have a single one with 20-byte entries. So, no need to support that.
